tests/test_packages/test_connections/test_p2p_libp2p_client/test_communication.py

```diff
@@ -622,3 +624,156 @@ def test_libp2pclientconnection_uri():
+        """Make an envelope for testing purposes."""
+        msg = DefaultMessage(
+            dialogue_reference=("", ""),
+            message_id=1,
+            target=0,
+            performative=DefaultMessage.Performative.BYTES,
+            content=b"hello",
+        )
+        envelope = Envelope(
+            to=receiver_address,
+            sender=sender_address,
+            protocol_specification_id=DefaultMessage.protocol_specification_id,
+            message=DefaultSerializer().encode(msg),
+        )
+        return envelope
+
+
+class TestLibp2pClientReconnectionSendEnvelope(BaseTestLibp2pClientReconnection):
+    """Test that connection will send envelope with error, and that reconnection fixes it."""
+
+    def test_envelope_sent(self):
+        """Test the envelope is routed."""
+        addr_1 = self.connection_client_1.address
+        addr_2 = self.connection_client_2.address
+        envelope = self._make_envelope(addr_1, addr_2)
+
+        # make the send to fail
+        with mock.patch.object(
+            self.connection_client_1.logger, "exception"
+        ) as _mock_logger:
+            # set the current writer to none, to raise error in send
+            self.connection_client_1._writer.close()
+            self.connection_client_1._writer = None
+            self.multiplexer_client_1.put(envelope)
+            delivered_envelope = self.multiplexer_client_2.get(block=True, timeout=20)
+            _mock_logger.assert_called_with(
+                "Exception raised on message send. Try reconnect and send again."
+            )
+
+        assert delivered_envelope is not None
+        assert delivered_envelope.to == envelope.to
+        assert delivered_envelope.sender == envelope.sender
+        assert (
+            delivered_envelope.protocol_specification_id
+            == envelope.protocol_specification_id
+        )
+        assert delivered_envelope.message == envelope.message
+
+
+@libp2p_log_on_failure_all
+class TestLibp2pClientReconnectionReceiveEnvelope(BaseTestLibp2pClientReconnection):
+    """Test that connection will receive envelope with error, and that reconnection fixes it."""
+
+    def test_envelope_received(self):
+        """Test the envelope is routed."""
+        sender = self.connection_client_2.address
+        receiver = self.connection_client_1.address
+        envelope = self._make_envelope(sender, receiver)
+
+        # make the receive to fail
+        with mock.patch.object(
+            self.connection_client_1.logger, "exception"
+        ) as _mock_logger, mock.patch.object(
+            self.connection_client_1._reader, "readexactly", side_effect=ConnectionError
+        ):
+            self.multiplexer_client_2.put(envelope)
+            delivered_envelope = self.multiplexer_client_1.get(block=True, timeout=20)
+            _mock_logger.assert_called_with(
+                RegexComparator(f"Connection error:.*. Try to reconnect and read again")
+            )
```

Comment:
@DavidMinarsch currently this test does not pass. The mock makes the reader descriptor of the connection to raise an error when reading. However, when reconnecting, the received data are lost. Is that the intended behaviour or a request for retransmission is missing?

Reply:
Ok, that's what I was worried about. The behaviour is not intended. We don't want any data to be lost. This will likely imply a go-side fix. Good to have the test case!

Reply:
I don't exclude improper mocking and testing, though.

Reply:
Yeah that's possible too. Although the above problem would match some of the reported issues.